### PR TITLE
borrowck: Prevent partial reinitialization of uninitialized structures

### LINKED ISSUE #18571: Undetected use after move
So I was curious how Drop handles cycles, because this could lead to dangling pointers, which is why `__del__` [might not be called in Python](https://docs.python.org/2/library/gc.html#gc.garbage). And indeed the below example does not call Drop.

But `t` is moved into `u`, so why does this compile at all?

```rust
struct Test {
    a: int,
    b: Option<Box<Test>>,
}

impl Drop for Test {
    fn drop(&mut self) {
        println!("Dropping {}", self.a);
    }
}

fn stuff() {
    let mut t = Test { a: 1, b: None};
    let mut u = Test { a: 2, b: Some(box t)};    
    t.b = Some(box u);
    println!("done");
}

fn main() {
    stuff();
    println!("Hello, world!")
}
```

### PR DESCRIPTION
**Pull request by Ryman**
This is a rebase of #18963

2nd commit is to accommodate @nikomatsakis's comments on the PR which handles his additional testcase. 
Will squash post-review if it's adequate.

Fixes #18571.

cc @pcwalton @nikomatsakis

### CODE REVIEW

File: src/librustc_borrowck/borrowck/check_loans.rs [hunk 1]
```diff
@@ -745,6 +745,26 @@ impl<'a, 'tcx> CheckLoanCtxt<'a, 'tcx> {
                                                      use_kind, lp_base);
             }
             LpExtend(ref lp_base, _, LpInterior(InteriorField(_))) => {
+                match lp_base.to_type().sty {
+                    ty::ty_struct(def_id, _) | ty::ty_enum(def_id, _) => {
+                        if ty::has_dtor(self.tcx(), def_id) {
+                            // In the case where the owner implements drop, then
+                            // the path must be initialized to prevent a case of
+                            // partial reinitialization
```
Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/check_loans.rs]: This code is an improvement on the prior PR that you rebased, thanks!

However, I do wonder whether we could simplify further and just call `self.check_if_path_is_moved(id, span, use_kind, lp_base);` after doing the `ty::has_dtor` check, the same way that the other cases below do.

Does that produce an error message that is too obscure?  (Or, is there some technical reason we cannot do that?)

**Comment by pnkfelix** [comment]
The tests look good to me.

@Ryman r=me after squashing (and I would like an answer to the question above if possible).

File: src/librustc_borrowck/borrowck/check_loans.rs (hunk 1 above)
Comment (Ryman) [inline src/librustc_borrowck/borrowck/check_loans.rs]: Off the top of my head, it was possible but the error was a more generic 'use of moved value `foo.bar`'.

I reasoned that the improved error message from the original changeset might be preferable but I'm willing to change it for simpler code.

Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/check_loans.rs]: How about this: Keep your code the way it is, but throw in a FIXME here and above `fn check_if_path_is_moved` noting that the body of `check_if_path_is_moved` is largely the same as the code you put into this function.  That way, if someone does this same thing again, maybe we will then be inspired to refactor the code over what error message to report.

(That, or you could add in such a refactoring yourself, as a followup commit post squash.)

**Comment by pnkfelix** [comment]
(the bors queue is close to empty right now, so I might just steal this and make a fresh PR for it.)

**Comment by pnkfelix** [comment]
closing in favor of #22219 (which just does the squashing); thanks for the contribution, I made sure to preserve your authorship on the squash.  :)

**Comment by Ryman** [comment]
@pnkfelix No worries, I didn't notice this before going ahead with a quick refactor, you can check https://github.com/Ryman/rust/commit/d57d5f36f0a62418ffe14cf90d37aebe05c1142e to see if it warrants pulling in or not. (The alternative, taking a closure didn't really remove much duplication)

**Pull request closed without merging**